src/Razor/src/Microsoft.VisualStudio.Editor.Razor/Snippets/XmlSnippetParser.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Composition;
+using System.Diagnostics;
+using System.IO;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.VisualStudio.Editor.Razor.Snippets;
+
+/// <summary>
+/// Server instance agnostic snippet parser and cache.
+/// This can be re-used across LSP servers as we're just storing an
+/// internal representation of an XML snippet.
+/// </summary>
+[Export(typeof(XmlSnippetParser)), Shared]
+internal partial class XmlSnippetParser
+{
+    private readonly ILogger? _logger;
+
+    [ImportingConstructor]
+    public XmlSnippetParser(ILoggerFactory? loggerFactory)
+    {
+        _logger = loggerFactory?.CreateLogger<XmlSnippetParser>();
+    }
+
+    /// <summary>
+    /// Cache to hold onto the parsed XML for a particular snippet.
+    /// </summary>
+    private readonly ConcurrentDictionary<string, ParsedXmlSnippet?> _parsedSnippetsCache = new();
```

Comment:
Is this file also copied and pasted? Because it is odd to me we have a cache here, and a `SnippetCache`, both with very different styles. Should `SnippetCache` just be a `ConcurentDictionary`? or will this cache end up never being used once `SnippetCache` is populated?

Reply:
Updated where we're caching things